chcemvediet/apps/inforequests/models/action.py

```diff
@@ -315,7 +315,7 @@ def previous_action(self):
 
     @cached_property
     def next_action(self):
-        return self.branch.action_set.order_by_created().after(self).first()
+        return self.branch.action_set.exclude(pk=self.pk).order_by_created().after(self).first()
```

Comment:
@martinmacko47 Vsimol som si, ze `next_action` sa sprava "zvlastne". Vo vela pripadoch vrati tu istu akciu. 
```
assert action_23.next_action != action_23  # AssertionError
``` 
Zatial sa nikde v kode nepouziva, takze netreba ani nic prerabat. Bude vsak napomocna pri zistovani ci je nejaka akcia posledna v branchi. Problem vznika pri `self.filter(Q(created__gt=other.created) | ...` tento filter zahrnie aj `other`, pri filtri `__lt` toto spravanie nenastava a metda `previous_action` pracuje spravne.

Reply:
@viliambalaz Nerozumiem, vies mi spravit minimalny example DB, kde tento problem nastava?

Ved `.filter(Q(created__lt=other.created) | Q(created=other.created, pk__lt=other.pk))` spravi select:
```
WHERE (
  "inforequests_action"."branch_id" = 2 AND (
    "inforequests_action"."created" > 2019-11-16 19:44:20.465456 OR (
      "inforequests_action"."id" > 3 AND
      "inforequests_action"."created" = 2019-11-16 19:44:20.465456
    )
  )
)
```
Cize selectnuta action ma `created` bud ostro vacsie nez povodna action, alebo ak je `created` rovne, tak `id` musi mat ostro vacsie. Cize ak aj obe akcie maju rovnaky cas vytvorenia, tak potom aj tak sa neselectne povodna akcia, kedze `id` musi byt ostro vacsie.

Alebo nebodaj podmienka `"inforequests_action"."created" > 2019-11-16 19:44:20.465456` je true aj ked su casy rovnake? Ak ano, tak potom zrejme nastava nejaka chyba pri zaokruhlovani casu niekde. Potom treba fixnut tu chybu. Workarround cez `.exclude(pk=self.pk)` iba skryje symptom v tom konkretnom pripade, ale bug ostane.

Reply:
@martinmacko47 Query mam rovnake.
>>> a26 = Action.objects.get(pk=26)
>>> a26
<Action: [26] Potvrdenie prijatia žiadosti>
>>> a26.next_action
<Action: [26] Potvrdenie prijatia žiadosti>
>>> a26.created
datetime.datetime(2019, 9, 3, 8, 42, 58, 486948, tzinfo=<django.utils.timezone.FixedOffset object at 0x7fd749cf8650>)
>>> Action.objects.filter(created=a26.created)
[]
>>> a26 in Action.objects.filter(created__lt=a26.created)
False
>>> a26 in Action.objects.filter(created__gt=a26.created)
True
>>> a26 in Action.objects.filter(created=a26.created)
False
>>> a26 in Action.objects.filter(created__lte=a26.created)
False
>>> a26 in Action.objects.filter(created__gte=a26.created)
True
>>> 
```
Tiez je zaujimave, ze to robi len pri niektorych instanciach action, nie pri vsetkych. Asi dane akcie ktore mi nefunguje si jednoducho vymazem z databazy a nebudem s nimi pracovat. Skusal som si online editor a ten Selectovat spravne.

Reply:
@viliambalaz Django v sqlite pri vytvoreni instancie ulozi datetime do DB ako `2020-07-03 10:46:13.391936` (bez timezone suffixu). Ale pri update tvaru `Action.objects.update(created=F(u'created') - datetime.timedelta(days=1))` ulozi do DB vyslednu hodnotu ako `2020-07-02 10:46:13.391936+00:00` (s timezone suffixom `+00:00`).

Hoci logicky `2020-07-02 10:46:13.391936+00:00` a `2020-07-02 10:46:13.391936` by malo byt to iste, v sqlite to nie je to iste, pretoze sqlite porovnava casy ako stringy. Takze `2020-07-02 10:46:13.391936+00:00` je vacsie ako `2020-07-02 10:46:13.391936`.

Bug je v [`_sqlite_format_dtdelta`](https://github.com/django/django/blob/1.7.4/django/db/backends/sqlite3/base.py#L569). V komentari tam pisu, ze `str(dt)` ma vratit datetime bez timezone suffixu, ale vrati ho s timezone suffixom. Lebo v [`typecast_timestamp`](https://github.com/django/django/blob/1.7.4/django/db/backends/sqlite3/base.py#L569) tam tu timezonu pridaju. Co som pozeral, tak to [fixli](https://github.com/django/django/commit/88637064b3118f90ae5a402e1b47a243ef88d656) az teraz nedavno pred pol rokom.

Jedine, co mi napada, co s tym spravit, je to monkey-pachnut. Tj. ze ak sme na localhoste, tak zobrat funkciu `_sqlite_format_dtdelta` a nahradit ju nasou funkciou, ktora bude vraciat stringy bez `+00:00` suffixu. A monkey-patch vyhodime, ked upgradneme Django na verziu, v ktorej je to uz fixnute.

Alternativne by sme mohli kaslat na cele sqlite a pouzivat aj pri vyvoji na localhoste MySQL. Aj tak pouzivat inu DB pri vyvoji, nez v produkcii, nie je uplne dobre. V takom pripade by sa vsak zislo napisat docker images, aby sme nemuseli instalovat MySQL lokalne a riesit konflikty medzi roznymi verziami MySQL pri roznych projektoch. V case, ked CHV vznikalo, tak docker este nebol.

Pripadne docasne monkey-pachnut `_sqlite_format_dtdelta`. A niekedy neskor prejst na MySQL.

Reply:
@viliambalaz Bug som vysunul to samostatnej issue: https://github.com/martinmacko47/chcemvediet/issues/293 Budeme to riesit v separatnom PR.